### PR TITLE
Fix FakeTensorProp on Module with Parameters or Buffers

### PR DESCRIPTION
**Pull request by wschin**
In `FakeTensorMode.__torch_dispatch__`, the output is now always computed by meta kernels in
```python
        try:
            with in_kernel_invocation_manager(self):
                r = func(*args, **kwargs)  # <----- "r" can be a real tensor.
        except NotImplementedError as not_implemented_error:
            # no meta kernel registered, fallback to kernel for the device
            if not self.allow_fallback_kernels:
                raise not_implemented_error
            return run_fallback_kernel(self, func, args, kwargs, not_implemented_error)

        return self.wrap_meta_outputs_with_default_device_logic(r, func, args, kwargs)
```
For example, I observed a CPU tensor is generated when executing `aten.addmm` when running `FakeTensorProp`. Therefore, I'd like to allow `FakeTensorMode` to wrap real tensor as `FakeTensor` during the computation. Does this PR look a good direction to fix this problem? If yes, I can go ahead and add some tests.

### CODE REVIEW

**Comment by eellison** [comment]
Yea, something doesn't look right here. +1 on repro

**Comment by wschin** [comment]
Looks like my code doesn't use FakeTensorMode properly. I have two FakeTensorMode's; one for converting model `Parameter`s to `FakeTensor`s and the other one is implicitly created in `FakeTensorProp` for shape inference.
```python
    fake_tensor_mode = FakeTensorMode()

    def to_fake_tensor(x):
        if isinstance(x, torch.Tensor) and not isinstance(x, FakeTensor):
            return fake_tensor_mode.from_tensor(x)
        return x

    fake_parameters_and_buffers = {
        k: to_fake_tensor(v)
        for k, v in itertools.chain(module.named_parameters(), module.named_buffers())
    }

    with stateless._reparametrize_module(
        decomposed_module, fake_parameters_and_buffers
    ):

    FakeTensorProp(decomposed_module).propagate(*args)
```
If I change
```python
 FakeTensorProp(decomposed_module).propagate(*args)
 ```
 to
 ```
  FakeTensorProp(decomposed_module, fake_tensor_mode).propagate(*args)
```
and use the passed-in `fake_tensor_mode` inside `FakeTensorProp`, I don't need this PR. Maybe the right fix is to allow `FakeTensorProp` to use custom `FakeTensorMode`. What do you think?

**Comment by ezyang** [comment]
Pass through sounds good to me.

**Comment by wschin** [comment]
Hi @ezyang, suggested fix is implemented in [f74d807](https://github.com/pytorch/pytorch/pull/88700/commits/f74d8072a2a26935e264e47a39466e58d681dc87) with a detailed test `test_fake_tensor_prop_on_nn_module` for explaining the reason and examining this fix.

File: test/test_fake_tensor.py [hunk 1]
```diff
@@ -663,5 +665,62 @@ def test_like_ops(self):
...
+                value = self.layer2(value)
+                return value
+
+        model = ToyNnModuleWithParameters()
+        value = torch.randn(5, 4)
+        # Convert nn.Module to GraphModule so that FakeTensorProp runs.
+        graph_model = torch.fx.symbolic_trace(model, (value,))
+        # The following block runs FakeTensorProp on graph_module w/to the same FakeTensorMode
+        #
+        # TODO(wschin): there should be an API to run FakeTensorProp for GraphModule
```
Comment (eellison) [inline test/test_fake_tensor.py:688]: The recommended way to do this is:
```
mode = FakeTensorMode()
with torch._subclasses.fake_tensor.FakeCopyMode(mode):
     mod_copied = copy.deepcopy(m)
mod_copied([mode.from_tensor(t) for t in inps])
```

**Review by eellison — APPROVED** [review]
lgtm - maybe wait for @ezyang as well

**Review by ezyang — APPROVED** [review]
Didn't review test

**Comment by wschin** [comment]
pytorchbot merge

**Pull request closed without merging**